src/lib/operators/sort.cpp

```diff
@@ -200,41 +243,72 @@ class Sort::SortImpl : public AbstractReadOnlyOperatorImpl {
       }
     }
 
-    // 3. Materialization of the result: We take the sorted ValueRowID Vector, create chunks fill them until they are
-    // full and create the next one. Each chunk is filled row by row.
-    auto materialization = std::make_shared<SortImplMaterializeOutput<SortColumnType>>(_table_in, _row_id_value_vector,
-                                                                                       _output_chunk_size);
-    auto output = materialization->execute();
-
-    const auto chunk_count = output->chunk_count();
-    for (auto chunk_id = ChunkID{0}; chunk_id < chunk_count; ++chunk_id) {
-      output->get_chunk(chunk_id)->set_ordered_by(std::make_pair(_column_id, _order_by_mode));
-    }
-
-    return output;
+    return _row_id_value_vector;
   }
 
+ protected:
   // completely materializes the sort column to create a vector of RowID-Value pairs
-  void _materialize_sort_column() {
+  void _materialize_sort_column(const std::shared_ptr<PosList>& pos_list) {
     auto& row_id_value_vector = *_row_id_value_vector;
     row_id_value_vector.reserve(_table_in->row_count());
 
     auto& null_value_rows = *_null_value_rows;
 
-    const auto chunk_count = _table_in->chunk_count();
-    for (ChunkID chunk_id{0}; chunk_id < chunk_count; ++chunk_id) {
-      const auto chunk = _table_in->get_chunk(chunk_id);
-      Assert(chunk, "Physically deleted chunk should not reach this point, see get_chunk / #1686.");
-
-      auto base_segment = chunk->get_segment(_column_id);
+    // If there was no PosList passed, this is the first sorting run and we simply fill our values and nulls data
+    // structures from our input table. Otherwise we will materialize according to the PosList which is the result of
+    // the last run.
+    if (pos_list) {
+      auto segment_ptr_and_accessor_by_chunk_id =
+          std::unordered_map<ChunkID, std::pair<std::shared_ptr<const BaseSegment>,
+                                                std::shared_ptr<AbstractSegmentAccessor<SortColumnType>>>>();
+      segment_ptr_and_accessor_by_chunk_id.reserve(_table_in->chunk_count());
+
+      // When there was a preceding sorting run, we materialize according to the passed PosList.
+      for (RowID row_id : *pos_list) {
+        const auto [chunk_id, chunk_offset] = row_id;
+
+        auto& segment_ptr_and_typed_ptr_pair = segment_ptr_and_accessor_by_chunk_id[chunk_id];
+        auto& base_segment = segment_ptr_and_typed_ptr_pair.first;
+        auto& accessor = segment_ptr_and_typed_ptr_pair.second;
```

Comment:
structured bindings?

Reply:
Done in 2e4a1e739bcc9286a783cba00336ca8e61325125